Description: don’t return invalid workspace rectangle
Origin: upstream, https://git.gnome.org/browse/libwnck/commit/?id=2fc0f058f00daf76
Last-Update: 2014-10-22

--- a/libwnck/pager.c
+++ b/libwnck/pager.c
@@ -788,6 +788,17 @@
 
   gtk_widget_get_allocation (widget, &allocation);
 
+  if (allocation.x < 0 || allocation.y < 0 ||
+      allocation.width < 0 || allocation.height < 0)
+    {
+      rect->x = 0;
+      rect->y = 0;
+      rect->width = 0;
+      rect->height = 0;
+
+      return;
+    }
+
   _wnck_pager_get_padding (pager, &padding);
   gtk_widget_style_get (widget,
 			"focus-line-width", &focus_width,
